src/traces/parcoords/base_plot.js

```diff
@@ -0,0 +1,80 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var d3 = require('d3');
+var Plots = require('../../plots/plots');
+var parcoordsPlot = require('./plot');
+var xmlnsNamespaces = require('../../constants/xmlns_namespaces');
+
+exports.name = 'parcoords';
+
+exports.attr = 'type';
+
+exports.plot = function(gd) {
+    var calcData = Plots.getSubplotCalcData(gd.calcdata, 'parcoords', 'parcoords');
+    if(calcData.length) parcoordsPlot(gd, calcData);
+};
+
+exports.clean = function(newFullData, newFullLayout, oldFullData, oldFullLayout) {
+    var hadParcoords = (oldFullLayout._has && oldFullLayout._has('parcoords'));
+    var hasParcoords = (newFullLayout._has && newFullLayout._has('parcoords'));
+
+    if(hadParcoords && !hasParcoords) {
+        oldFullLayout._paperdiv.selectAll('.parcoords-line-layers').remove();
+        oldFullLayout._paperdiv.selectAll('.parcoords-line-layers').remove();
+        oldFullLayout._paperdiv.selectAll('.parcoords').remove();
+        oldFullLayout._paperdiv.selectAll('.parcoords').remove();
+        oldFullLayout._glimages.selectAll('*').remove();
+    }
+};
+
+exports.toSVG = function(gd) {
+
+    var imageRoot = gd._fullLayout._glimages;
+    var root = d3.selectAll('.svg-container');
+    var canvases = root.filter(function(d, i) {return i === 0;})
+        .selectAll('.parcoords-lines.context, .parcoords-lines.focus');
+
+    function canvasToImage() {
+        var canvas = this;
+        var rect = canvas.getBoundingClientRect();
```

Comment:
This is a good point: we should definitely minimize our use of `getBoundingClientRect`, and it might be worthwhile to look through the places we use it and see which ones we can remove and which others we may need to alter.

The one you point to in `drawing` @rreusser is a case we really do need it - there's no other way AFAIK to get the dimensions and positioning of text, but it's important HOW it's invoked: on a test element attached directly to `<body>`, ie not part of `gd`. That way we can guarantee that it's attached to the DOM and displayed, which is not always the case with `gd` itself when the plot is first rendered - imagine creating a bunch of plots in different tabs of your app, most of which are hidden initially.

I think we can say that during interactions `getBoundingClientRect` is safe on elements within `gd` (it has to be visible to be interacted with, right? I guess in principle not if an app developer triggers something like a hover effect manually...) but a quick search shows we're probably using it in unsafe ways during initial rendering.